rfcs/000-ml-feature-discovery.md

```diff
@@ -0,0 +1,183 @@
+### Advanced Search Page
+![Table Lineage Preview](../assets/036/adv-search.png)
+
+### Feature Details Page
+##### Sample Tab
+![Table Lineage Preview](../assets/036/sample-tab.png)
+##### Statistics Tab
+![Table Lineage Preview](../assets/036/stats-tab.png)
+##### Upstream Tab
+![Table Lineage Preview](../assets/036/upstream-tab.png)
+##### Generation Code Tab
+![Table Lineage Preview](../assets/036/generation-tab.png)
+
+## Reference-level Explanation (aka Technical Details)
+
+### Amundsen Databuilder
+The databuilder library will be expanded to include a feature Model that includes the following:
+#### Nodes
+- Feature node:
+    - Key
+    - Name
+    - Version
+    - Feature group
+    - Entity
+    - Data type
+    - Last updated
+#### Relationships
+Here we show the relationships between feature and other nodes, as well as the metadata that this represents: 
+- Feature <FEATURE/FEATURE_OF> Database
+    - Availability
+- Feature <COLUMN/COLUMN_OF> Column <> Badge
+    - Partition column
+    - Note: this column key will have feature key in key instead of a table key
+- Feature <DESCRIPTION/DESCRIPTION_OF> Description
+- Feature <DESCRIPTION/DESCRIPTION_OF> Programmatic_Description
+- Feature <QUERY_OF/HAS_QUERY> Query
+    - Code snippet/SQL: relating it to query from dashboard because we need an external link too
+- Feature <STAT/STAT_OF> Stat
+- Feature <HAS_BADGE/BADGE_FOR> Badge
+    - contains PII
+- Feature <OWNER_TAG/TAGGED_BY_OWNER> Tag
+    - Owner Tags, tag type can now be “owner”, not just default
+- Feature <TAG/TAGGED_BY> Tag
+    - Amundsen User Tags
+- Feature <OWNER/OWNER_OF> User
+- Feature <WATERMARK> Watermark
+    - Date range
+
+### Amundsen Common
+Amundsen common will have the folllowing schema which represents the feature response between servies:
+```
+class Feature:
+   key: str  # ex: test_feature_group_name/test_feature_name/1.2.0
+   name: str
+   version: str  # ex: "1.2.0"
+   status: str
+   feature_group: str
+   entity: Optional[List[str]]
+   data_type: Optional[str]
+   availability: List[str]
+   description: Optional[str]
+   owners: List[User]
+   badges: List[Badge]
+   owner_tags: Optional[List[Tag]] # non editable
+   tags: List[Tag] # editable
+   generation_code: Optional[Query]  # reuse Query entity from Dashboards
+   programmatic_descriptions: List[ProgrammaticDescription] 
+   watermarks: List[Watermark]  # date range
+   stats: List[Stat]
+   last_updated_timestamp: Optional[int]
+   created_timestamp: int
+   partition_column: Optional[Column]
+ 
+class FeatureSummary:
+   key: str  # ex: test_feature_group_name/test_feature_name/1.2.0
+   name: str
+   version: str  # ex: "1.2.0"
+   availability: List[str]
+   entity: Optional[List[str]]
+   description: Optional[str]
+```
+### Amundsen Metadata
+The features API methods will basically be identical to that for tables, but the response will follow the outline of the common schema specified above.
+#### Requests
+- GET feature details
```

Comment:
each of them would have their own page right now given that we don't have a comprehensive design for history of a dataset, feature, dashboard, etc.

Reply:
After thinking through this a bit more, I really like the definition of feature groups. It really resonates with how I've worked on teams in the past that associate features to the pipeline that creates them, the entity they represent, etc. From a UX perspective, I think displaying features at the feature group instead of having one page for each feature could be better (if you _had_ to map this to today's world feature group = table, feature = column).

The thought here is that the highest impact discoverability will occur immediately after a user finds a feature that is relevant to them. The feature group would inherently contain similar features and help augment the discoverability of features.

Thoughts?

Reply:
@sewardgw, you are absolutely right about feature groups when conceptualizing them as feature group = table and feature = column. A large # of features have been derived from Hive tables historically. 

However, we are increasingly witnessing online features derived from streaming sources. There this concept of feature groups break down. Features may not be generated via a single pipeline and may not be contained inside a single encapsulating container such as a table. Displaying such features has to be done without a concept of a feature group. 

Therefore, feature group is an optional metadata. And we decided to display features as independent entity. However, if one wishes to look at all features corresponding to a given feature's feature group, we will enable that. Let me know if this makes sense.

Reply:
Makes sense! Love the thought around having options for iterations on this feature to continue the buildout if needed.